src/transformers/generation/watermarking.py

```diff
@@ -62,76 +68,78 @@ class WatermarkDetectorOutput(ModelOutput):
 
     >>> # first generate text with watermark and without
-    >>> args = {"bias": 2.5, "context_width": 3, "seeding_scheme": "selfhash", "greenlist_ratio": 0.25, "hashing_key": 15485863}
-    >>> out_watermarked = model.generate(**inputs, watermarking_args=args, do_sample=False, max_length=20)
+    >>> watermarking_config = WatermarkingConfig(bias=2.5, seeding_scheme="selfhash")
+    >>> out_watermarked = model.generate(**inputs, watermarking_config=watermarking_config, do_sample=False, max_length=20)
     >>> out = model.generate(**inputs, do_sample=False, max_length=20)
 
     >>> # now we can instantiate the detector and check the generated text
-    >>> detector = WatermarkDetector(bos_token_id=tok.bos_token_id, vocab_size=tok.vocab_size, device="cuda:0", watermarking_args=args)
-    >>> detection_out_watermarked = detector(out_watermarked[:, input_len:], return_dict=True)
-    >>> detection_out = detector(out[:, input_len:], return_dict=True)
-    >>> detection_out_watermarked["prediction"]
+    >>> detector = WatermarkDetector(model_config=model.config, device="cpu", watermarking_config=watermarking_config)
+    >>> detection_out_watermarked = detector(out_watermarked, return_dict=True)
+    >>> detection_out = detector(out, return_dict=True)
+    >>> detection_out_watermarked.prediction
     array([ True,  True])
 
-    >>> detection_out["prediction"]
-    array([ False,  False])
+    >>> detection_out.prediction
+    array([False,  False])
     ```
     """
 
     def __init__(
         self,
-        bos_token_id: int,
+        model_config: PretrainedConfig,
         device: str,
-        vocab_size: int,
-        watermarking_args: Dict,
+        watermarking_config: Union[WatermarkingConfig, Dict],
         ignore_repeated_ngrams: bool = False,
+        max_size: int = 128,
     ):
-        self.bos_token_id = bos_token_id
+        if isinstance(watermarking_config, WatermarkingConfig):
+            watermarking_config = watermarking_config.to_dict()
+
+        self.bos_token_id = (
+            model_config.bos_token_id if not model_config.is_encoder_decoder else model_config.decoder_start_token_id
+        )
+        self.greenlist_ratio = watermarking_config["greenlist_ratio"]
         self.ignore_repeated_ngrams = ignore_repeated_ngrams
-        self.greenlist_ratio = watermarking_args["greenlist_ratio"]
-        self.processor = WatermarkLogitsProcessor(vocab_size=vocab_size, device=device, **watermarking_args)
+        self.processor = WatermarkLogitsProcessor(
+            vocab_size=model_config.vocab_size, device=device, **watermarking_config
+        )
+        self._get_ngram_score_cached = lru_cache(maxsize=max_size)(self._get_ngram_score)
 
-    @lru_cache(maxsize=2**32)
-    def _get_ngram_score_cached(self, prefix: torch.LongTensor, target: int):
+    def _get_ngram_score(self, prefix: torch.LongTensor, target: int):
         """Expensive re-seeding and sampling is cached."""
```

Comment:
This should be a comment next to the application of `lru_cache`, so we know why we're applying a cache :)